### ISSUE TITLE
async trait for runtime

### ISSUE THREAD

**Issue by NobodyXu** [nrc/portable-interoperable#13]
I propose to add a new trait for the runtime, which provides portable APIs for creating sockets, pipe, `AsyncFd`, timers and etc.

### Motivation

As mentioned in [this comment](https://github.com/nrc/portable-interoperable/issues/12):

> I don't think having async IO traits like Read/Write/BufRead is enough to unify the ecosystem.
> 
> Many crates, like reqwest, http server, [openssh-mux-client](https://github.com/openssh-rust/openssh-mux-client) need to create a network socket (tcp/udp/unix socket) and without a way to create it in a portable manner, they will resort back to use a specific runtime.
> 
> There is also crates like [tokio-pipe](https://github.com/yskszk63/tokio-pipe) which wraps the pipe for tokio users, I think we need to have some way to create [AsyncFd](https://docs.rs/tokio/latest/tokio/io/unix/struct.AsyncFd.html) in a portable manner.

### Details

Here is a rough scratch of what I expected, many APIs are elided because there are too many APIs:

(NOTE that these APIs are heavily influenced by `tokio` because it is the most familiar async runtime API to me)

```rust
use std::io::Result;

trait RuntimeHandle: Clone {
    type JoinHandle<T>: Future<Output = T>;

    fn block_on(&self, f: impl Future<Output = T>) -> T;

    fn spawn(&self, f: impl Future<Output = T> + Send + Sync + 'static) -> Self::JoinHandle<Result<T>>;
    fn spawn_local(&self, f: impl Future<T> + 'static) -> Self::JoinHandle<Result<T>>;

    /// Provided function, automatically call spawn if f is Send + Sync, otherwise call spawn_local.
    /// 
    /// # Motivation
    /// 
    /// Suppose that a user want to write some data into TcpStream at background,
    /// they will likely to use spawn.
    /// 
    /// However, spawn requires Send + Sync, in order to use it, they would have to write
    /// `RuntimeNet<TcpStream: Send + Sync>`, which is cumbersome.
    /// 
    /// Another option is to just use spawn_local, but it only gives concurrency, not
    /// parallelism.
    /// 
    /// Thus, Runtime should have a provided method spawn_auto to automatically
    /// spawn it on local or spawn it on any thread.
    /// 
    /// This will make the portable Runtime trait much easier to use, and
    /// it would even improve the portability, since users do not have to worry
    /// about the "Send + Sync" bound anymore.
    fn spawn_auto(&self, f: impl Future<T> + 'static) -> Self::JoinHandle<Result<T>>;
}

trait Runtime {
    type Handle: RuntimeHandle;

    async fn yield_now();

    fn block_in_place<F: FnOnce() -> R, R>(f: F) -> R;

    fn handle(&self) -> &Self::Handle;

    // true for io_uring, false for epoll, used to determine whether file APIs should just use `spawn_blocking`
    fn supports_non_blocking_file_read_write(&self) -> bool;
}

/// Provides AsyncHandle
trait RuntimeAsyncHandle: Runtime {
    type AsyncHandle: AsyncHandle;

    #[cfg(unix)]
    fn from_fd(fd: std::os::unix::io::OwnedFd) -> Result<Self::AsyncHandle>;

    #[cfg(windows)]
    fn from_handle(fd: std::os::windows::io::OwnedHandle) -> Result<Self::AsyncHandle>;
}

trait AsyncHandle: Read + Write + Seek + Clone {
   type ReadBuffered: AsyncHandle + BufRead;
   type WriteBuffered: AsyncHandle + BufRead;

    fn into_buffered_read(self, n: usize) -> Self::ReadBuffered;
    fn into_buffered_write(self, n: usize) -> Self::WriteBuffered;
}

/// Net ops, optional
trait AsyncHandleNet: AsyncHandle {
    async fn accept(&self) -> Result<AsyncHandle>;
    fn into_accept_stream(self) -> impl Stream<Item = Result<AsyncHandle>>;

    async fn connect(&self, ...) -> Result<()>;

    async fn sendmsg(&self, buffer: IoSlice<'_>, aux: Auxiliary) -> Result<usize>;
    async fn recvmsg(&self, buffer: IoSliceMut<'_>, aux: Auxiliary) -> Result<usize>;

    fn into_recvmsg_stream(self) -> impl Stream<Item = Result<Buffer, Auxiliary>>;
}

/// File ops, optional
trait AsyncHandleFile: AsyncHandle {
    async fn fsync(&self) -> Result<()>;

    /// If the runtime/os does not support this, then it can issue fsync instead.
    async fn fdatasync(&self) -> Result<()>;

    /// If the runtime/os does not support this, then it can issue fdatasync instead.
    async fn sync_file_range(&self, off: u64, n: u64) -> Result<()>;

    /// If the runtime/os does not support this, then it can use read + write to simulate this.
    async fn copy_file_range(&self, off_in: u64, other: & impl AsyncHandle, off_out: u64, len: u64) -> Result<()>;

    async fn metadata(&self) -> Result<MetaData>;
    async fn set_metadata(&self, metadata: MetaData) -> Result<()>;
}

/// fs operations, like open, optional
trait RuntimeFs: Runtime {
    type File: Read + BufRead + Write + BufWrite + Seek;

    async fn open(&self) -> Result<Self::File>;
}

/// signal handling, optional
trait RuntimeSignal: Runtime {
    fn ctrl_c(&self) -> Result<()>;
}

/// spawn process, optional
trait RuntimeProcess: Runtime {
    type Command: ...;

    fn command(&self) -> Self::Command;
}

/// Perhaps the same set of methods as tokio::time::Interval?
trait Interval {...}

/// time handling, optional but most likely implemented
trait RuntimeTime: Runtime {
    type Interval: Interval;

    fn interval_at(start: Option<Instant>, period: Duration) -> Self::Interval;
}

#[cfg(unix)]
mod unix {
    use super::*;

    enum Madvice { ... }
    trait RuntimeMem: Runtime {
        async unsafe fn madvice(&self, addr: *mut (), len: usize, advice: Madvice) -> Result<()>;
    }

    trait RuntimeAsyncHandleExt: RuntimeAsyncHandle {
        async fn openat2(&self, ...) -> Result<Self::AsyncHandle>;
    }

    trait AsyncHandlePipeExt: AsyncHandle {
        async fn splice(&self, ...) -> Result<usize>;
        async fn tee(&self, ...) -> Result<usize>;
    }

    trait AsyncHandleNetExt: AsyncHandleNet {
        async fn sendfile(&self, ...) -> Result<usize>;
    }

    enum SignalKind { ... }
    trait RuntimeSignalExt: RuntimeSignal {
        type Signal: Stream;

        fn signal(kind: SinglaKind) -> Result<Self::Signal>;
    }
}
```

**Comment by SabrinaJewson** [comment]
It would be simpler to not have separate process/net/file/stdio APIs but rather provide a single unified `Socket` type (actually, might be better to have `Fd` on Unix and `Socket` + `Handle` on Windows) similar to socket2’s API as well as a `spawn_blocking` function, and then `TcpStream`, `File`, `Process` etc can all build on that generically. So, more like:

```rs
pub trait Runtime {
    #[cfg(unix)]
    type Fd: Fd;

    // true for io_uring, false for epoll, used to determine whether file APIs should just use `spawn_blocking`
    fn supports_non_socket_fds(&self) -> bool;
}

// unix-only
pub trait Fd: From<OwnedFd> + Into<OwnedFd> + AsyncRead + AsyncWrite {}
```

and `std` can define a `TcpStream<Inner>`, where on Unix `Inner: Fd` and on Windows `Inner: Socket`.

**Comment by SabrinaJewson** [comment]
Ah, you misunderstood me. I meant that process/net/file/stdio should all be together, but time and task spawning should still be separate.

**Comment by NobodyXu** [comment]
> Ah, you misunderstood me. I meant that process/net/file/stdio should all be together, but time and task spawning should still be separate.

Sorry for that, integrating these traits does sound reasonable, but I am still not completely sure this is a good idea though.

**Comment by NobodyXu** [comment]
Oops...

**Comment by NobodyXu** [comment]
Having this method:

```
    // true for io_uring, false for epoll, used to determine whether file APIs should just use `spawn_blocking`
    fn supports_non_socket_fds(&self) -> bool;
```

in `Runtime` will definitely be necessary.

**Comment by NobodyXu** [comment]
Also, we need to find a way to obtain the runtime.

I personally think the proposal of having a "context" is the best solution for this, since it is zero-cost (does not require thread-local variable) and the crate can easily put a bound on the "context".

**Comment by NobodyXu** [comment]
> Ah, you misunderstood me. I meant that process/net/file/stdio should all be together, but time and task spawning should still be separate.

After a second thought, I think it will be better to keep it separated.

While it's true that process/net/file/stdio can be put together, the implementation of the runtime often split them into multiple features.

For example, [tokio](https://docs.rs/crate/tokio/latest/features) provides feature flags for each one of them to speedup compilation and reduce bloat.

Similarly, [async-std](https://docs.rs/crate/async-std/latest/features) also provides feature flags.

Thus, I think it will be better to leave them as separate traits so that the runtime can choose to provide feature flags for each of them.

**Comment by SabrinaJewson** [comment]
Under my design, types like TcpStream, process::Child, Stdin etc would still be equally feature flagged. All that wouldn't be feature flagged would be the underlying low-level system call wrappers which would be common between the high level net/process/io interfaces anyway.

**Comment by NobodyXu** [comment]
> Under my design, types like TcpStream, process::Child, Stdin etc would still be equally feature flagged. All that wouldn't be feature flagged would be the underlying low-level system call wrappers which would be common between the high level net/process/io interfaces anyway.

I am a bit confused.

Do you mean that process/net/file/stdio would be in the same trait be feature gated?

**Comment by NobodyXu** [comment]
This actually sounds good.

I think we need the runtime to implement AsyncHandle (can be constructed from OwnedFd or windows's OwnedHandle), Socket for async accept implementation, process spawning and timer.

**Comment by NobodyXu** [comment]
@SabrinaJewson I've adopted your suggestion and updated the proposal.

**Comment by NobodyXu** [comment]
@nrc I have written down a detailed proposal and it seems that we need at least type GAT for the spawning tasks.

**Comment by NobodyXu** [comment]
# Amendment Proposal: Archive zero-cost abstraction

## Motivation

Currently, in order to implement the portable `Runtime` trait, the async runtime
must use reference counting (`Rc`/`Arc`) or global `static` variables.

None of the solutions above is zero-cost.

Reference counting requires boxing and every time a new `AsyncHandle` and other types
or `Future` is created, it needs to increase the reference counting and it also
needs to decrease it on `Drop`.

Global `static` variable is also not zero-cost since it might makes the executable
larger or requires use of `once_cell`.

Thus, I decided to propose this to archive zero-cost abstraction.

## Modification

In order to archive zero-cost, we must eliminate the need of reference counting
or global `static` variables and replace them with a reference to the runtime.

To do so, we must add lifetime to every type that implements
the portable runtime trait.

There is two way to accomplish this, one is to use lifetime GAT and the other one is
to have a new rust language feature that provides `'self` lifetime.

### Lifetime GAT

We would need to apply the following modifications:

```rust
trait Handle: Clone {
    type JoinHandle<'a, T>: JoinHandle<T>;
}
trait RuntimeAsyncHandle: Runtime {
    type AsyncHandle<'a>: AsyncHandle;
}
trait AsyncHandle: Read + Write + Seek {
    type ReadBuffered<'a>: AsyncHandle + BufRead;
    type WriteBuffered<'a>: AsyncHandle + BufWrite;
    type ReadWriteBuffered<'a>: AsyncHandle + BufRead + BufWrite;
}
trait AsyncHandleNet: AsyncHandle {
    type AcceptStream<'a>: Stream<Item = Result<AsyncHandle>>;
    type RecvMsgStream<'a>: Stream<Item = Result<(Buffer, Auxiliary)>>;
}
trait RuntimeProcess: Runtime {
    type Command<'a>: ...;
}
trait RuntimeTime: Runtime {
    type Interval<'a>: Interval;
}
#[cfg(unix)]
mod unix {
    trait RuntimeSignalExt: RuntimeSignal {
        type Signal<'a>: Stream<Item = Result<SigInfo>>;
    }
}
```

This requires each type in the traits to be modified and add `<'a>` bound
to them.

This also means that we would have to depend on GAT and it might further
complicate the portable async story, especially for new users.

### `'self` lifetime

On the contrary, using `'self` lifetime is trival.

No modification to `triat`, we just need to add `'self` to the type specified
when implementing these traits.

### `spawn`

In additional to these changes, we also need to relax the `'static` lifetime bound
in `spawn`:

```rust
fn spawn<'this, 'future: 'this, F, T>(&'this self, f: F)
    where F: Future<Output = T> + Send + Sync + 'future -> Self::JoinHandle<T>;
```

This makes sure that the futures' lifetime is at least as long as the runtime.

Though the implementation of runtime will be more complicated, as they have to handle the order of dropping correctly,
manually create the vtable for the future since they cannot use `dyn Future<...>` anymore
and they also need to `Pin` their runtime to ensure that it cannot forgotten safely.

## Expected implementation

I would expect the following implementation:

```rust
struct SampleRtInner {
    ...
}

struct SampleRt {
    /// Declare thread_pool first to ensure it will
    /// be dropped before inner.
    thread_pool: ScopedThreadPool,
    inner: SampleRtInner,
}

impl Runtime for Pin<&SampleRt> {
    ...
}

impl SampleRt {
    pub fn new() -> Self {
        todo!()
    }

    pub fn new_scoped<F, T>(f: F) -> T
        where F: FnOnce(Pin<&Self>) -> T
    {
        let this = Self::new();
        std::pin::pin!(this);

        f(this.deref())
    }
}
```

## Call for review

I know this amendment proposal must be controversial as this requires either
GAT or a new feature `'self`, but I think if we want truely portable and zero-cost
runtime traits, it is the way to go.

@nrc @SabrinaJewson Can you guys review this please?
Thanks in advance!

**Comment by nrc** [comment]
Hey @NobodyXu I haven't looked in detail at the amendment above, I'm still at the information and requirements gathering stage of work on executors. But two comments: I expect using lifetime GATs is fine in general, they seem on a solid path towards stabilisation. I don't think being zero-cost in the sense of having no runtime overhead at all is necessary, the cost of a few statics or an occasional reference count incr/decr is well worth it for the sake of improved ergonomics (if that is in fact the trade-off here).

**Comment by NobodyXu** [comment]
> I expect using lifetime GATs is fine in general, they seem on a solid path towards stabilisation.

It's good to hear that!

> the cost of a few statics or an occasional reference count incr/decr is well worth it for the sake of improved ergonomics (if that is in fact the trade-off here).

Well, I don't think using a few statics or reference counting improve ergonomics.

The problem here is zero-cost runtime abstraction requires lifetime GAT, but if that is fine, then I don't think it will have a lot impact on ergonomics.

In fact, I would argue the current API makes more sense.

Users of this portable runtime can clearly see how their `AsyncHandle`, `Interval` and futures are created and understand it instead of treating it like black magic and get confused when it failed.

For example, many people using `tokio` might attempted to create `AsyncFd` or `TcpSocket` in non-async context, like a regular `fn f() -> TcpSocket`.

When using that function `f`, it will work fine most of the time, but it is easy for beginners to abuse and call that in non-async place where `#[tokio::main]`/`#[tokio::test]` has not been used or a `tokio::runtime::Runtime` has been created but `Runtim::{enter, block_on}` is not called.

In that case, the code will compile, but fail at runtime with a panic.

With this portable runtime proposal, this won't happen, because users need a `Runtime` trait (probably as context) to create `AsyncHandle`, `TcpSocket` or etc and it will fail at compile time if they forget to add that as a context.

**Comment by NobodyXu** [comment]
> the cost of a few statics or an occasional reference count incr/decr is well worth it for the sake of improved ergonomics (if that is in fact the trade-off here).

In additional to the ergonomic improvement I mentioned in the last comment, the amendment proposal also improves ergonomic for `spawn`.

Currently, `spawn` has to be `'static`, which is understandable but too strict.
With this amendment proposal, `spawn` can accept any future that lives as long as the runtime itself.

**Comment by NobodyXu** [comment]
@nrc The original proposal also uses specialisation:

```rust
// Provided function, automatically call spawn if f is Send + Sync, otherwise call spawn_local.
fn spawn_auto<T>(&self, f: impl Future<Output = T> + 'static) -> Self::JoinHandle<T>;
```

### Motivation

Suppose that a user want to write some data into TcpStream at background,
they will likely to use spawn.

However, spawn requires Send + Sync, in order to use it, they would have to write
`RuntimeNet<TcpStream: Send + Sync>`, which is cumbersome.

Another option is to just use spawn_local, but it only gives concurrency, not
parallelism.

Thus, Runtime should have a provided method spawn_auto to automatically
spawn it on local or spawn it on any thread.

This will make the portable Runtime trait much easier to use, and
it would even improve the portability, since users do not have to worry
about the "Send + Sync" bound anymore.

**Comment by NobodyXu** [comment]
This proposal definitely needs more feedbacks, espeically from whom maintains the async runtime.

@rust-lang/wg-async, @rust-lang/libs-api
@Darksonn @joshtriplett  @ihciah @Noah-Kennedy @HippoBaro

**Comment by Darksonn** [comment]
Well, I looked it over and it seems rather complicated. But perhaps that's unavoidable. I don't think the current location of your `spawn_local` method is compatible with the `spawn_local` in Tokio due to the requirement of using a `LocalSet`, which does not have a handle type.

**Comment by NobodyXu** [comment]
> Well, I looked it over and it seems rather complicated.

Part of that is because I also want to support file/fs async operation for io-uring based async executor.

> the `spawn_local` in Tokio due to the requirement of using a `LocalSet`

Is there any reason why `tokio` has to use a `LocalSet`?
[`async_std::task::spawn_local`](https://docs.rs/async-std/latest/async_std/task/fn.spawn_local.html) does not require anything equivalent to a `LocalSet` AFAIK.

**Comment by Darksonn** [comment]
Well, on the multi-threaded executor, local spawns are a problem due to our [`block_in_place`](https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html) method. The `block_in_place` method is supposed to move everything on the current worker thread to a new worker thread, which is impossible if you can `spawn_local` on that runtime.

As for the current-thread runtime that doesn't support `block_in_place`, it's because the `Runtime` object itself is `Send`, so moving the `Runtime` would let you move the non-Send task across threads.

**Comment by NobodyXu** [comment]
You are right, that is something I didn't take into account.
I've removed `block_in_place` API from my proposal.

**Comment by Noah-Kennedy** [comment]
Unfortunately, uring-based runtimes are going to necessitate very different APIs for their completion-based operations.

**Comment by Noah-Kennedy** [comment]
That's one difference, another is that (currently working on doing this for tokio-uring) there are a lot of neat flags you can set on the SQE pre-submission that a user may well want to configure, including things like provided buffers, fixed buffers, etc.

**Comment by NobodyXu** [comment]
@Noah-Kennedy For the difference between read and write, you might want to readn and give feedback for it on #5 

> there are a lot of neat flags you can set on the SQE pre-submission that a user may well want to configure, including things like provided buffers, fixed buffers, etc.

Portable runtime traits are designed mainly for library, so perhaps the user can configure provided/fixed buffers can be preconfigured that in `main` or something like that, where they know which runtime they are using?

We can also add functions as hints or adjust the buffer of `BufRead`/`BufWrite`:

```rust
    fn into_buffered_read(self, n: usize) -> Self::ReadBuffered;
    fn into_buffered_write(self, n: usize) -> Self::WriteBuffered;

    fn into_buffered(self, read_buffer_size: usize, write_buffer_size: usize) -> Self::ReadWriteBuffered;
```

We might not able to expose all functionalities io-uring provides, so I think we should focus on providing a portable, common interface for crates/library writers.

For these cases where more details can significantly improve performance, perhaps they can use specialisation to archive this?

**Comment by joshtriplett** [comment]
I am *incredibly* confused by the level of complexity here.

Supplying a runtime doesn't mean supplying file or socket types or handling for fds or handles or process creation or similar. The standard library will already have all of those, and they'll be the *same* types and functions no matter what runtime you're using; that's a big part of what it means to unify the ecosystem. If you want to use the types supplied by some other ecosystem, you can reference that ecosystem's crates. We're going to have *one* implementation of AsyncFile, not an abstraction that lets each runtime have its own.

The runtime trait should support spawning futures (`spawn` and `spawn_blocking`), and *possibly* `spawn_local` (though I'm not sure if we need that or if we'll end up with a better solution in terms of "futures that can be sent before they're polled but not after").

The runtime trait shouldn't require an associated type for JoinHandle; we should have a single type for that, with unified semantics, so that callers can work with one without having to abstract over it using `dyn` or similar. We can store enough information inside that JoinHandle that the runtime can get the information it needs back out. I would expect that storing a pointer should suffice for that.

**Comment by joshtriplett** [comment]
Sketch:

```rust
trait Runtime {
    fn block_on<T>(&self, f: impl Future<Output = T>) -> T;
    fn spawn<T>(&self, f: impl Future<Output = T> + Send + 'static) -> JoinHandle<T>;
    fn spawn_local<T>(&self, f: impl Future<Output = T> + 'static) -> JoinHandle<T>;
}

trait RuntimeBlocking: Runtime {
    fn spawn_blocking<T>(&self, f: impl Future<Output = T> + 'static) -> JoinHandle<T>;
}
```

Note that `spawn_blocking` should take a future as well, so that it can make both sync and async calls without having to `block_on` for the latter.

**Comment by ibraheemdev** [comment]
And JoinHandle holds a type-erased pointer + vtable?

**Comment by Noah-Kennedy** [comment]
@joshtriplett how exactly would we be able to have a single, runtime-independent type for e.g an async file or socket? I'm not sure how this would work due to reliance on the IO driver. Would we need to standardize on an interface for an IO driver as well then, and just move the standardization issue further down the stack?

**Comment by jyn514** [comment]
I wish that `AsyncFd` and `AsyncRead` were standard. I think if we had those too, it would be much much easier to interoperate, even without standardized *types*.

**Comment by Noah-Kennedy** [comment]
@jyn514 I don't think you can standardize AsyncFd terribly well given that the thing it's an attempt to standardize isn't even cross-platform.

**Comment by joshtriplett** [comment]
@ibraheemdev I don't think we need the vtable to be per-handle; we could just have it be per-runtime, in which case we'd probably need a way to register it as part of registering a runtime.

**Comment by joshtriplett** [comment]
@Noah-Kennedy If we *can't* have a runtime-independent type for an async file or socket, we might as well just put a full runtime in std; I'd sooner do that than make every program deal with an abstraction over the type of `AsyncFile`. I'm very much taking it as a given that we'll have a single `std::fs::AsyncFile` (module bikeshedding the name and module path); that's part of the value of standardization.

Roughly what I had in mind:

- An `Async` type that wraps anything implementing `AsFd` (on UNIX) or `AsHandle` (on Windows), for convenience of implementation.
- Extensions to Waker and similar to provide an easy way to say "I don't need to be woken up until this thing is readable" or "until this thing is writable". This should give enough information to allow the runtime to turn that into an epoll or WaitForMultipleObjects or similar.
- Implementations of everything *other* than `File` would be trivial wrappers around `Async` (or possibly even just type aliases for it along with `impl Async<...> { ... }` blocks adding type-specific methods; there are tradeoffs there that we should consider).
- The implementation of `File` would use `spawn_blocking`.
- If in the future we have another implementation based on `io_uring`, it'd start out as an external crate, and might eventually become a new backend for std.

**Comment by Noah-Kennedy** [comment]
@joshtriplett I don't think locking the async ecosystem into the development pace of std is a good idea. Having a minimal set of traits outside of which you need to specialize is certainly doable and probably the best approach. Having a single async runtime will create a massive amount of fragmentation, as there are many legitimate reasons you may need your own runtime. We are better off implementing these things as traits rather than concrete types.

**Comment by ibraheemdev** [comment]
With an advanced enough contexts and capabilities proposal, the standard library have something like `struct File<with T: AsyncRead + AsyncWrite + ...> { inner: T }` that lets us keep the minimal traits and have better ergonomics.

**Comment by Noah-Kennedy** [comment]
That could also work quite well

**Comment by NobodyXu** [comment]
> @Noah-Kennedy If we _can't_ have a runtime-independent type for an async file or socket, we might as well just put a full runtime in std; I'd sooner do that than make every program deal with an abstraction over the type of `AsyncFile`. I'm very much taking it as a given that we'll have a single `std::fs::AsyncFile` (module bikeshedding the name and module path); that's part of the value of standardization.
> 
> Roughly what I had in mind:
> 
>     * An `Async` type that wraps anything implementing `AsFd` (on UNIX) or `AsHandle` (on Windows), for convenience of implementation.
> 
>     * Extensions to Waker and similar to provide an easy way to say "I don't need to be woken up until this thing is readable" or "until this thing is writable". This should give enough information to allow the runtime to turn that into an epoll or WaitForMultipleObjects or similar.
> 
>     * Implementations of everything _other_ than `File` would be trivial wrappers around `Async` (or possibly even just type aliases for it along with `impl Async<...> { ... }` blocks adding type-specific methods; there are tradeoffs there that we should consider).
> 
>     * The implementation of `File` would use `spawn_blocking`.
> 
>     * If in the future we have another implementation based on `io_uring`, it'd start out as an external crate, and might eventually become a new backend for std.
> 
>     * For platforms that have _neither_ file descriptors nor handles, we'd need to know more about the operation of those platforms to know how to support async on them.

@joshtriplett This is exactly what I am trying to propose here.

Each runtime just need to implement `AsyncHandle` to provide basic functionalities like read/write/seek for any fd on unix/handle on windows.

There are also extension traits such as `AsyncHandleNet` for runtime that provides such functionalities and `AsyncHandleFile` and `RuntimeFs` for performing async file/fs io for io-uring based runtime.

Then the portable runtime trait would just expose a wrapper over these functionalities and provides types like `TcpStream` and `UnixStream` if `AsyncHandleNet` is implemented, `File` and other fs I/O types if `AsyncHandleFile` and `RuntimeFs` are implemented, otherwise just a default implementation using `spawn_blocking`.

The runtime only needs to expose set of APIs for dealing with `AsyncHandle` and the portable runtime trait would provide a set of highlevel, easy-to-use APIs.

(P.S. This idea is proposed by @SabrinaJewson )

`RuntimeProcess` is required because waiting for process to exit in an async manner is also very platform dependent.

Same for `RuntimeSignal`, it is platform dependent.

**Comment by Noah-Kennedy** [comment]
@NobodyXu I think the approach here is overall the best way of doing this, however we do need to be careful with the design here so as to not "cut off" certain runtime strategies from being able to meet the standard.

**Comment by joshtriplett** [comment]
> I don't think locking the async ecosystem into the development pace of std is a good idea. Having a minimal set of traits outside of which you need to specialize is certainly doable and probably the best approach. Having a single async runtime will create a massive amount of fragmentation, as there are many legitimate reasons you may need your own runtime. We are better off implementing these things as traits rather than concrete types.

I'm not arguing that we *should* do it that way; I mentioned it as an alternative to having to use traits to abstract over *this much* of a runtime. I'd prefer to do neither of those, and instead just have concrete types for almost everything except the runtime itself.

**Comment by joshtriplett** [comment]
> @joshtriplett This is exactly what I am trying to propose here.
> 
> Each runtime just need to implement `AsyncHandle` to provide basic functionalities like read/write/seek for any fd on unix/handle on windows.

Why is that in any way runtime-specific, beyond "hey runtime, here's a file descriptor / handle, don't wake my task until it's X-able"?

std provides platform abstractions for a variety of platforms. We don't implement the synchronous `File` by having an abstract interface that a platform-specific crate can implement to provide the backend for `File`; we just implement `File` on each platform we support. We should do the same for async.

> There are also extension traits such as `AsyncHandleNet` for runtime that provides such functionalities and `AsyncHandleFile` and `RuntimeFs` for performing async file/fs io for io-uring based runtime.
>
> Then the portable runtime trait would just expose a wrapper over these functionalities and provides types like `TcpStream` and `UnixStream` if `AsyncHandleNet` is implemented, `File` and other fs I/O types if `AsyncHandleFile` and `RuntimeFs` are implemented, otherwise just a default implementation using `spawn_blocking`.

Why is any of that runtime-specific?

Also, as with the `AsyncRead` and `AsyncWrite` traits, I think we'd be *much* better off providing a mechanism that works for non-`io_uring` and *separately* considering what interface would be best for `io_uring`, given how different its requirements are.

> `RuntimeProcess` is required because waiting for process to exit in an async manner is also very platform dependent.

But not runtime-dependent. std provides synchronous processes, and can also provide asynchronous processes.

> Same for `RuntimeSignal`, it is platform dependent.

But not runtime-dependent.

**Comment by Noah-Kennedy** [comment]
So, as an example, you can do readiness via io_uring with IORING_OP_POLL_ADD, which has certain advantages over epoll for readiness-based IO. This is something up to the runtime, to the point where this is probably going to be used inside tokio once it gets io_uring support in the future.

**Comment by joshtriplett** [comment]
@Noah-Kennedy You absolutely can, and I'm suggesting that we shouldn't design a single interface that covers every *possible* runtime at the expense of complexity, rather than an interface that supports *many* runtimes. I'm suggesting that for the same reason that we're not designing I/O traits (AsyncRead and AsyncWrite) to cover both epoll and io_uring, we're designing separate traits for those.

**Comment by ibraheemdev** [comment]
Another example, tokio relies on edge-triggered notifications, meaning it has to do an extra step of updating internal state (`clear_readiness`) when an I/O operation returns `WouldBlock`. If all you have is `poll_ready(fd: RawFd)` there's no way of doing that.

**Comment by joshtriplett** [comment]
@ibraheemdev I wasn't expecting a completely unmodified version of any existing runtime to drop into the standard library's runtime trait with zero adaptation or modification. I'd expect it to be more likely that there's a `runtime_x_stdruntime` crate adapting each `runtime_x` to the standard library interface.

**Comment by ibraheemdev** [comment]
How about [we move this to zulip](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async/topic/async.20trait.20for.20runtime).

**Comment by Noah-Kennedy** [comment]
If we want non-generic types, we should do this with vtables for registering readiness, or, in the case of completion-based APIs, for submitting and receiving events. This gives us as much flexibility as possible.

**Comment by joshtriplett** [comment]
> This gives us as much flexibility as possible.

This is a requirement I'd like to question. :slightly_smiling_face:

Also, I'm starting to wonder if the term "runtime" is possibly too abstract, and leading to too general of a solution. I'm wondering in particular if it'd be clearer to talk separately about task spawning and about a reactor; I'm not sure if we want to be abstracting fully over the latter.

**Comment by NobodyXu** [comment]
> > @joshtriplett This is exactly what I am trying to propose here.
> > Each runtime just need to implement `AsyncHandle` to provide basic functionalities like read/write/seek for any fd on unix/handle on windows.
> 
> Why is that in any way runtime-specific, beyond "hey runtime, here's a file descriptor / handle associated with this task/waker, don't wake my task until it's X-able"? The runtime doesn't need to support read/write/seek, it just needs to support "when do I wake up this task". The async runtime's job is to spawn and wake tasks, not to do I/O.
> 
> (Ideally I'd invert the interface: instead of "hey runtime, here's a file descriptor / handle", I'd go "here's a way to find out if there's a file descriptor / handle associated with this task/waker" and then the runtime can check that. But that's more of an implementation detail.)
> 
> std provides platform abstractions for a variety of platforms. We don't implement the synchronous `File` by having an abstract interface that a platform-specific crate can implement to provide the backend for `File`; we just implement `File` on each platform we support. We should do the same for async.

As others had replied, I added these so that both polling and completion based system like io-uring can be efficiently supported.

For example, when reading a file using io-uring, you won't use polling interface (and `epoll` doesn't work anyway).
Completion based interface are far more efficient and you are able to submit a batch of them in the same syscall.
 
> > There are also extension traits such as `AsyncHandleNet` for runtime that provides such functionalities and `AsyncHandleFile` and `RuntimeFs` for performing async file/fs io for io-uring based runtime.
> > Then the portable runtime trait would just expose a wrapper over these functionalities and provides types like `TcpStream` and `UnixStream` if `AsyncHandleNet` is implemented, `File` and other fs I/O types if `AsyncHandleFile` and `RuntimeFs` are implemented, otherwise just a default implementation using `spawn_blocking`.
> 
> Why is any of that runtime-specific?
> 
> Also, as with the `AsyncRead` and `AsyncWrite` traits, I think we'd be _much_ better off providing a mechanism that works for non-`io_uring` and _separately_ considering what interface would be best for `io_uring`, given how different its requirements are.

Same as above, for efficiency reason.

> > `RuntimeProcess` is required because waiting for process to exit in an async manner is also very platform dependent.
> 
> But not runtime-dependent. std provides synchronous processes, and can also provide asynchronous processes.

That requires std to have some internal signal handle registered that might run at background to modify some global variables on some unix targets, or old version of Linux.

Then there are often more efficient alternatives, such as using signalfd, or PidFd on Linux.

And it is totally possible for some other targets to come up with their own efficient implementation.

And then you need to distribute that signal to specific future waiting on specific pid.
For PidFd approach, this is trivial.

For signal based approach, this is very complex and should not be in std.

> > Same for `RuntimeSignal`, it is platform dependent.
> 
> But not runtime-dependent.

Maybe we can do this in std, given that the traditional signal handling and signalfd are the only way to do this and there are no niche target that provides more efficient implementation.

**Comment by NobodyXu** [comment]
> > This gives us as much flexibility as possible.
> 
> This is a requirement I'd like to question. 🙂
> 
> Also, I'm starting to wonder if the term "runtime" is possibly too abstract, and leading to too general of a solution. I'm wondering in particular if it'd be clearer to talk separately about task spawning and about a reactor; I'm not sure if we want to be abstracting fully over the latter.

For me, it is not just flexibility, but also efficiency.

This portable runtime trait is designed for library crates to adopt so that they can write efficient softwares without having a hard dependency on tokio/async-std.

Currently, a lot of library crates like reqwest, axiom, etc either have hard dependency on tokio, or have some feature gates to switch between tokio and async-std.

That is exactly what this trait is trying to fix.

If it is not efficient enough when adopting a new runtime, such as io-uring based ones, and they still need to resort to using specific runtime, then it would completely void the point of this.

**Comment by Noah-Kennedy** [comment]
I agree completely with @NobodyXu here regarding both efficiency and supporting different implementation strategies, which **needs to be viewed as a hard requirement here**.

**Comment by NobodyXu** [comment]
> If we want non-generic types, we should do this with vtables for registering readiness, or, in the case of completion-based APIs, for submitting and receiving events. This gives us as much flexibility as possible.

It will be great if we can archive that, but it might be quite hard as we have to manually poll things instead of having simple `async fn`.

**Comment by Noah-Kennedy** [comment]
I was thinking you fire off a callback to tell the reactor "register me to be woken if X fd is read/write ready", for instance with readiness-based events.

**Comment by NobodyXu** [comment]
> I was thinking you fire off a callback to tell the reactor "register me to be woken if X fd is read/write ready", for instance with readiness-based events.

That seems to be reasonable.

**Comment by joshtriplett** [comment]
Summarizing and expanding on some discussion from Zulip: @Noah-Kennedy explained some of the details of tokio's implementation, and I explained some of the details of async-std's implementation, and the key difference that became clear is that tokio's approach requires tight integration between the reactor (handling I/O readiness) and the executor (running tasks) to interleave the two, while async-std's approach allows separating the two entirely. One of many tradeoffs between those approaches is whether our abstraction can assume separability of those two concerns, or whether it needs to allow for the two to have dependencies on each other.

(In the description below, I am *not* going to weigh in on which approach has higher performance, whether by throughput or latency or any other consideration, on an absolute scale. I'm just going to discuss how different approaches would affect each model relative to itself. Also, in each case I'm going to try to state both advantages and disadvantages.)

I think those tradeoffs directly inform what approach we want to use here:
- We could assume separability, and have a very simple executor abstraction and either a simple reactor abstraction (separate from the executor) or no reactor abstraction (and our own reactor). This would also make I/O types (File, TcpStream, etc) simpler. This would still allow plugging in any executor, and when using async-std as the executor this would not add any overhead, but when using tokio as the executor this would add overhead (in the form of a separate reactor or partial reactor) compared to using tokio "natively".
- We could allow for non-separability, and provide a complex executor/reactor joint abstraction. This would allow tokio-style runtimes to have a "more native" implementation matching what they would normally do if not abstracted. This would add complexity and overhead to async-std-style runtimes. This would mean a much "wider" interface between std and the executor/reactor, which means more to design and stabilize, and more to potentially get wrong.
- We could have our own reactor modeled after the tokio approach, while having an executor abstraction. (I already covered the possibility of having our own reactor modeled after the async-std approach in the first bullet, since that approach can treat the reactor as separable.) This would require expanding the executor abstraction enough to support hooks for the reactor, and would require having more tuning parameters for our reactor.
- We could have our own executor and reactor, modeled after either approach. (Not necessarily a preferred alternative; mentioning it for completeness only.)

**Comment by joshtriplett** [comment]
I do think there's an additional consideration that comes into play here. *Because* this is std, and we have stricter stability requirements than the rest of the ecosystem, that gives somewhat more weight and downside to API surface area, and augurs in favor of simpler interfaces/abstractions. That *doesn't* mean we can't consider more complex interfaces, if we buy sufficient benefits with that; it just means there need to be more upsides to outweigh the downsides.

**Comment by NobodyXu** [comment]
I would like executor and reactor to be separated, but the executor should provide APIs so that async runtime such as tokio can use it.

IMHO this would make it more compositable and might even make it possible for `rayon` and `tokio` or other async ractor to share the same thread pool.

The key problem with this approach is to design the set of APIs for `tokio`.

**Comment by Noah-Kennedy** [comment]
@joshtriplett at the same time, the stability cuts the other way, if we make many common runtime designs non-interoperable with std async types, we have probably harmed the ecosystem more than we have helped it.

**Comment by NobodyXu** [comment]
Thank you @joshtriplett @Noah-Kennedy @ibraheemdev for providing me with feedback!

# async portable runtime 0.2.0

In portable runtime 0.2.0, I split the functionalities into executor and reactor to make the interface simpler, more compositable and easier to standardize.

This does not mean that all runtimes have to support plugable executor.
Runtime like `tokio` can still implement both executor and reactor for efficiency.

Portable runtime 0.2.0 also avoids use of `async` and generics in the reactor APIs so that the users of this crate can simply take a `&dyn Reactor` via context.

## Executor

Below is the APIs for the executor, it is significantly simplified:

```rust
trait Executor {
    type JoinHandle<'a, T>: JoinHandle<T>;

    async fn yield_now(&self);

    fn block_on<F, T>(&self, f: F) -> T
    where
        F: Future<Output = T>;

    /// If there are local tasks, then the executor can refuse to block.
    fn try_block_in_place<F, R>(&self, f: F) -> Result<R, F>
    where
        F: FnOnce() -> R;

    fn spawn<'this, 'future: 'this, F, T>(&'this self, f: F) -> Self::JoinHandle<'_, T>
    where
        F: Future<Output = T> + Send + 'future;

    fn spawn_local<'this, 'future: 'this, F, T>(&'this self, f: F) -> Self::JoinHandle<'_, T>
    where
        F: Future<Output = T> + 'future;

    /// Provided function, automatically call spawn if f is Send, otherwise call spawn_local.
    ///
    /// # Motivation
    ///
    /// Suppose that a user want to write some data into TcpStream at background,
    /// they will likely to use spawn.
    ///
    /// However, spawn requires Send, in order to use it, they would have to write
    /// `RuntimeNet<TcpStream: Send>`, which is cumbersome.
    ///
    /// Another option is to just use spawn_local, but it only gives concurrency, not
    /// parallelism.
    ///
    /// Thus, Runtime should have a provided method spawn_auto to automatically
    /// spawn it on local or spawn it on any thread.
    ///
    /// This will make the portable Runtime trait much easier to use, and
    /// it would even improve the portability, since users do not have to worry
    /// about the "Send" bound anymore.
    fn spawn_auto<'this, 'future: 'this, F, T>(&'this self, f: F) -> Self::JoinHandle<'_, T>
    where
        F: Future<Output = T> + 'future;

    fn spawn_blocking<'this, 'f: 'this, F, R>(&'this self, f: F) -> Self::JoinHandle<'_, R>
    where
        F: FnOnce() -> R + Send + 'f,
        R: Send + 'f;
}

enum JoinError {
    /// Out of memor when spawning
    Oom,
    /// The future panic
    Panic(Box<dyn Any + Send>),
    /// Cancelled
    Cancelled,
    /// Other
    Other(Box<dyn error::Error + Send>),
}
trait JoinHandle<T>: Future<Output = Result<T, JoinError>> {
    fn cancel(&self);
    fn is_done(&self) -> bool;
}
```

In additional to the APIs described above, there is also `ExecutorExt` that provides additional APIs for the reactor:

```rust
use std::error;
use std::io;
use std::num;

/// Extension trait for executor/reactor co-op, optional and experimental.
trait ExecutorExt: Executor {
    /// Can only register once
    fn register_poller<'this, 'f: 'this, Poller>(&'this self, poller: Poller) -> Result<(), Poller>
    where
        Poller: Fn(std::thread::ThreadId) -> io::Result<Action> + Send + Sync + 'f;

    /// Can only register once.
    fn register_callback<'this, 'f: 'this, G, CallBack>(
        &self,
        callback_generator: G,
    ) -> Result<(), G>
    where
        G: Fn(std::thread::ThreadId) -> CallBack + Send + Sync + 'f,
        CallBack: FnMut(Event<'_>) -> Action;
}

enum Event<'a> {
    /// Future is executed and return `Poll::Pending`.
    FuturePending(&'a FutureStatistics),
    /// Future is executed and completed.
    FutureDone(FutureId),
    /// Tasks are migrated to current thread.
    TaskMigrated(num::NonZeroUsize),
    /// Tasks are taken away from current thread.
    TasksTakenAway(num::NonZeroUsize),
}

enum Action {
    /// Steal tasks from other threads.
    StealTask,
    /// Gift tasks from other threads.
    GiftTask,
    /// Run poller.
    Poll,
    /// Continue without any action.
    Noop,
}

/// Must uniquely identify every future
#[derive(Copy, Clone, Debug)]
struct FutureId(u64);

/// Can obtain statistics like how long a future has been running,
/// how many times it yeilds with `Poll::Pending`.
struct FutureStatistics {
    future_id: FutureId,
    // ...
}
```

## Reactor

Below is the basic interface for the reactor:

```rust
use std::io;
use std::marker;
use std::task::{Context, Poll};

trait ReactorPoller {
    fn register(&self, pollable: Pollable<'_>, interest: Interest) -> io::Result<Registration<'_>>;
    fn deregister(&self, pollable: Pollable<'_>, registration: Registration) -> io::Result<()>;

    fn poll_read_ready(
        &self,
        cx: &mut Context,
        registration: &mut Registration<'_>,
    ) -> Poll<io::Result<()>>;
    fn poll_write_ready(
        &self,
        cx: &mut Context,
        registration: &mut Registration<'_>,
    ) -> Poll<io::Result<()>>;

    fn clear_read_ready(&self, registration: &mut Registration<'_>);
    fn clear_write_ready(&self, registration: &mut Registration<'_>);
}

/// An opaque type holding platform-dependent fd/handle
struct Pollable<'a>;

impl<'a> Pollable<'a> {
    #[cfg(unix)]
    pub fn from_fd(fd: std::os::unix::io::BorrowedFd<'a>) -> Self;

    #[cfg(windows)]
    pub fn from_handle(handle: std::os::windows::io::BorrowedHandle<'a>) -> Self;

    #[cfg(windows)]
    pub fn from_socket(socket: std::os::windows::io::BorrowedSocket<'a>) -> Self;
}

/// An opaque type.
struct Registration<'reactor>([usize; 2], marker::PhantomData<&'reactor ()>);

/// A bitflag that contains READABLE and WRITEABLE.
struct Interest;
```

`ReactorTime`:

```rust
use std::time;

trait ReactorTimer {
    fn poll_sleep(&self, duration: time::Duration, cx: &mut Context) -> Poll<()>;
}
```

`ReactorCompletion` for completion based runtime like `tokio-uring`:

```rust
trait ReactorCompletion {
    fn poll_for_response(
        &self,
        pollable: &RequestDesc<'_>,
        cx: &mut Context,
    ) -> Poll<io::Result<()>>;
    fn get_response(&self, pollable: RequestDesc<'_>) -> Option<RawResponse<'_>>;
    fn consume_response(&self, raw_response: RawResponse<'_>) -> Poll<io::Result<()>>;

    fn register_pollable(
        &self,
        pollable: Pollable<'_>,
        cx: &mut Context,
    ) -> Poll<io::Result<Request<'_, PollableDesc<'_>>>>;
    fn unregister_pollable(
        &self,
        pollable_desc: PollableDesc<'_>,
        cx: &mut Context,
    ) -> Poll<io::Result<Request<'_, ()>>>;

    fn register_buffer(
        &self,
        buffer: OwnedBuffer,
        cx: &mut Context,
    ) -> Poll<io::Result<Request<'_, BufDesc<'_>>>>;
    fn unregister_buffer(
        &self,
        buf_desc: BufDesc<'_>,
        cx: &mut Context,
    ) -> Poll<io::Result<Request<'_, ()>>>;

    fn accept(
        &self,
        pollable_desc: &PollableDesc<'_>,
        cx: &mut Context,
    ) -> Poll<io::Result<Request<'_, OwnedPollable>>>;
    fn register_multishot_accept(
        &self,
        pollable_desc: &PollableDesc<'_>,
        cx: &mut Context,
    ) -> Poll<io::Result<MultishotRequest<'_, OwnedPollable>>>;

    unsafe fn read(
        &self,
        pollable_desc: &PollableDesc<'_>,
        buf_desc: &mut BufDesc<'_>,
        cx: &mut Context,
    ) -> Poll<io::Result<Request<'_, usize>>>;
    unsafe fn write(
        &self,
        pollable_desc: &PollableDesc<'_>,
        buf_desc: &BufDesc<'_>,
        cx: &mut Context,
    ) -> Poll<io::Result<Request<'_, usize>>>;
}

/// An opaque type
struct RequestDesc<'reactor>(u64, marker::PhantomData<&'reactor ()>);

struct Request<'reactor, T>(
    &'reactor dyn ReactorCompletion,
    Option<RequestDesc<'reactor>>,
    PhantomData<T>,
);

impl<'reactor, T> Request<'reactor, T> {
    pub unsafe fn new<R: ReactorCompletion>(
        reactor: &'reactor R,
        request_desc: RequestDesc,
    ) -> Self;
}
impl<T> Future for Request<'_, T>
where
    T: FromRawResponse,
{
    // poll_for_response => get_response => FromRawResponse::from => consume_response
}

struct MultishotRequest<'reactor, T>(RequestDesc<'reactor>, marker::PhantomData<T>);
impl<'reactor, T> MultishotRequest<'reactor, T> {
    pub fn request(&self) -> Request<'reactor, T>;
}

struct RawResponse<'reactor> {
    cqe_addr: *const (),
    index: usize,
    phantom: marker::PhantomData<&'reactor ()>,
}

trait FromRawResponse {
    unsafe fn from(raw_response: &RawResponse<'reactor>) -> Self;
}

/// Similar to Pollable but owned.
struct OwnedPollable;

/// An opaque type
struct PollableDesc<'reactor>(u64, marker::PhantomData<&'reactor ()>);

struct OwnedBuffer(*mut u8, usize);
impl OwnedBuffer {
    unsafe fn new(ptr: *mut u8, len: usize) -> Self {
        Self(ptr, len)
    }
}

/// An opaque type
struct BufDesc<'reactor>(u64, marker::PhantomData<&'reactor ()>);
```

**Comment by NobodyXu** [comment]
## async executor 0.3

After posting about async executor 0.2, I realized it is possible to design the APIs of executor without any generics, `async` or associated item.

And I also realized that the `ExecutorExt` API is a bit too complex so I simplified them a bit and also removed the use of generics.

### `Executor`

```rust
use std::error;
use std::task::{Context, Poll};

trait Executor {
    /// Always return `Poll::Pending` then wake up the task.
    fn yield_now(&self, cx: &mut Context<'_>) -> Poll<!>;

    /// If there are local tasks, then the executor can refuse to block.
    fn try_enter_blocking_mode(&self) -> Result<(), ()>;

    /// Fail if not in blocking mode.
    fn try_exit_blocking_mode(&self) -> Result<(), ()>;

    /// This new API leaves how to allocate the future, how to implement
    /// the JoinHandle and the managment of lifetime to the users of this
    /// API.
    ///
    /// # Safety
    ///
    /// future must live until it is cancelled or done.
    unsafe fn spawn(&self, future: SpawnableFuture) -> Result<(), SpawnError>;

    /// # Safety
    ///
    /// future must live until it is cancelled or done.
    unsafe fn spawn_local(&self, future: SpawnableLocalFuture) -> Result<(), SpawnError>;

    /// # Safety
    ///
    /// function must live until it is done.
    unsafe fn spawn_blocking(&self, fn_once: SpawnableFnOnce) -> Result<(), SpawnError>;
}

enum SpawnError {
    /// Out of memor when spawning
    Oom,
    /// Other
    Other(Box<dyn error::Error>),
}

struct FutureVtable {
    /// Must not panic
    poll: fn(*mut (), cx: &mut Context<'_>) -> Poll<()>,
    /// Must not panic
    drop: fn(*mut ()),
}

struct SpawnableFuture {
    /// Must be `Send`able
    data: *mut (),
    vtable: FutureVtable,
}
unsafe impl Send for SpawnableFuture {}

impl Unpin for SpawnableFuture {}
impl Future for SpawnableFuture {
    type Output = ();

    // Call vtable.poll
}

impl Drop for SpawnableFuture {
    // Call vtable.drop
}

struct SpawnableLocalFuture {
    data: *mut (),
    vtable: FutureVtable,
}

impl Unpin for SpawnableLocalFuture {}
impl Future for SpawnableLocalFuture {
    type Output = ();

    // Call vtable.poll
}

impl Drop for SpawnableLocalFuture {
    // Call vtable.drop
}

struct FnOnceVtable {
    /// Must not panic
    call: fn(*mut ()),
    /// Must not panic
    drop: fn(*mut ()),
}

struct SpawnableFnOnce {
    /// Must be `Send`able
    data: *mut (),
    vtable: FnOnceVtable,
}
unsafe impl Send for SpawnableFnOnce {}

impl FnOnce for SpawnableFnOnce {
    // Call vtable.call
}

impl Drop for SpawnableFnOnce {
    // Call vtable.drop
}
```

### `ExecutorExt`

```rust
use std::io;
use std::marker;
use std::num;

/// Extension trait for executor/reactor co-op, optional and experimental.
trait ExecutorExt: Executor {
    /// Can register multiple hanlder generator.
    ///
    /// On every thread, executor will call the generator to create
    /// a per-thread event handler.
    ///
    /// When there are multiple per-thread handlers, they will be
    /// called in arbitary order.
    ///
    /// # Safety
    ///
    /// event_handler_generator must live until it is unregistered or this
    /// executor is dropped.
    ///
    /// The event handler returned by the event_handler_generator must also
    /// live as long as event_handler_generator.
    unsafe fn register_event_handler(
        &self,
        event_handler_generator: EventHandlerGenerator,
    ) -> EventHandlerId<'_>;

    /// This will unregister the event handler generator and all event handler
    /// it generated.
    fn unregister_event_handler(
        &self,
        event_handler_id: EventHandlerId<'_>,
    ) -> EventHandlerGenerator;
}

struct EventHandlerGeneratorVtable {
    /// Must not panic
    generate_new_handle:
        fn(*const (), thread_id: std::thread::ThreadId) -> io::Result<EventHandler<'_>>,
    /// Must not panic
    drop: fn(*mut ()),
}
struct EventHandlerGenerator {
    /// Must implement `Send` and `Sync`
    data: *mut (),
    vtable: EventHandlerGeneratorVtable,
}
unsafe impl Send for EventHandlerGenerator {}
unsafe impl Sync for EventHandlerGenerator {}

impl Fn for EventHandlerGenerator {
    // Call vtable.generate_new_handle
}
impl Drop for EventHandlerGenerator {
    // Call vtable.drop
}

struct EventHandlerVtable {
    /// Must not panic
    handle: fn(*mut (), event: Event<'_>) -> io::Result<Action>,
    /// Must not panic
    drop: fn(*mut ()),
}
struct EventHandler<'event_handler_generator> {
    data: *mut (),
    vtable: EventHandlerVtable,
    phantom: PhantomData<&'event_handler_generator EventHandlerGenerator>,
}

impl FnMut for EventHandler {
    // Call vtable.handle
}
impl Drop for EventHandler {
    // Call vtable.drop
}

/// Opaque type
struct EventHandlerId<'executor>(u64, marker::PhantomData<&'a ()>);

enum Event<'a> {
    /// Future is executed and return `Poll::Pending`.
    FuturePending(&'a FutureStatistics),
    /// Future is executed and completed.
    FutureDone(FutureId),
    /// Tasks are migrated to current thread.
    TaskMigrated(num::NonZeroUsize),
    /// Tasks are taken away from current thread.
    TasksTakenAway(num::NonZeroUsize),
}

enum Action {
    /// Steal tasks from other threads.
    StealTask,
    /// Gift tasks to other threads.
    GiftTask,
    /// Continue without any action.
    Noop,
}

/// Must uniquely identify every future
#[derive(Copy, Clone, Debug)]
struct FutureId(u64);

/// Can obtain statistics like how long a future has been running,
/// how many times it yeilds with `Poll::Pending`.
struct FutureStatistics {
    future_id: FutureId,
    // ...
}
```

**Comment by Noah-Kennedy** [comment]
So I've taken some time to look through this again.

Regarding this approach, I don't think this gets us much.

First, starting with the spawn methods, having them be unsafe like this makes the whole abstraction useless. If the purpose is for library authors to be able to _not_ care what runtime they are on, then it is self-defeating to make these methods unsafe and tied to whatever lifetimes the runtime supports, as the library authors don't know this information. These methods should simply require `'static` to fit the base case that all runtimes support.

Second, I think we are taking the wrong approach here. Currently, we seem to be trying to define the behavior that runtimes should adopt in a very tight-fitting manner, and I don't think we actually solve any problems here. Library users don't care if there is work stealing, and they don't really benefit from an API that stubs out a reactor like this, as this is an implementation detail of the runtimes that should be encapsulated from them.

I think what we really need here is to have common traits for different types, and to have a factory object of some sort which allows us to construct types matching those traits via it's associated types. This is more similar to the initial proposal. This allows library authors to write code independent of the runtime they are on, so long as that runtime offers a factory that they can use. It would be quite simple actually from an authors perspective, especially as compared to the other options discussed.

**Comment by NobodyXu** [comment]
> First, starting with the spawn methods, having them be unsafe like this makes the whole abstraction useless. If the purpose is for library authors to be able to not care what runtime they are on, then it is self-defeating to make these methods unsafe and tied to whatever lifetimes the runtime supports, as the library authors don't know this information. These methods should simply require 'static to fit the base case that all runtimes support.

The std library would provide wrappers over these `unsafe` primitives like `std::task::yield_now`, `std::task::spawn`, `std::task::spawn_local`, `std::task::spawn_auto`, `std::task::spawn_blocking`, `std::task::try_block_in_place`.

The std library would implement the `JoinHandle` that is portable to any runtime and the executor only needs to execute these tasks in any possible manner.

With the existing `unsafe` API, it is also possible to implement scoped spawning in std, though I don't mind making the exsiting spawning API `safe` and then add a new set of scoped spawning APIs to `Executor`.

> Second, I think we are taking the wrong approach here. Currently, we seem to be trying to define the behavior that runtimes should adopt in a very tight-fitting manner, and I don't think we actually solve any problems here. Library users don't care if there is work stealing, and they don't really benefit from an API that stubs out a reactor like this, as this is an implementation detail of the runtimes that should be encapsulated from them.

I presume you are talking about `ExecutorExt` here.

It is not a trait exposed to ordinary users, but rather the reactor implementation so that they can register their own poller to the executor and co-op with any executor, inspired by @joshtriplett when discussing on zulip.

Of course, this is kind of my wet dream since its APIs are indeed too tight, so I will probably remove it for now.

> I think what we really need here is to have common traits for different types, and to have a factory object of some sort which allows us to construct types matching those traits via it's associated types. This is more similar to the initial proposal. This allows library authors to write code independent of the runtime they are on, so long as that runtime offers a factory that they can use. It would be quite simple actually from an authors perspective, especially as compared to the other options discussed.

I actually changed that to this because of concerns raised by @joshtriplett .

The previous one uses generics and associated items (directly or through use of `async` in trait), so it is impossible to use it as a trait object and requires monomorphization in every `async` crate which might be too slow and prevents sharing the code if multiple reactors are used in the same binary somehow.

@joshtriplett also mentioned that I have the completion APIs packaged into the existing polling APIs, making it harder to change the completion APIs and add more APIs in the future.

With these concerns in mind, I decided to make the new reactor a low-level API that does not use generics or associated items at all, and split the completion and polling APIs into `ReactorCompletion` and `ReactorPoller`.

Then we can have a high-level API wrapping these lowlevel `ReactorPoller`/`ReactorCompletion` and it is actually simpler to implement.

For example, `ReactorPoller` only has 6 methods, which corresponds to `tokio::io::unix::AsyncFd`'s methods.
The std library can use it to implement its own `AsyncHandle`, then uses that to implement `TcpSocket`, `UnixSocket` and etc using existing facilities in the std library.

That will remove a lot of duplicate code to deal with `TcpSocket`, `UnixSocket` and etc in `tokio` and `async-std`.

For `ReactorCompletion`, it models closely after io-uring's actual interface so it is also pretty straight-forward to implement for `tokio-uring` and then we will again provide high-level APIs similar to it.

The high-level APIs can use specialisation to detect presence of `ReactorCompletion` like this:

```rust
trait Reactor: ReactorPoller + Sealed {
    fn to_reactor_completion(&self) -> Option<&dyn ReactorCompletion> {
        None
    }
}
impl<T: ReactorPoller + ReactorCompletion + Sealed> Reactor for T {
    fn to_reactor_completion(&self) -> Option<&dyn ReactorCompletion> {
        Some(self)
    }
}
```

which can be used in generic functions or as a trait object, then we would have highlevel APIs like this:

```rust
struct TcpStream {
    inner: std::net::TcpStream,
    /// Registration for ReactorPoller
    registration: Registration,
}
impl TcpStream {
    pub fn connect<A: ToSocketAddrs>(addr: A) -> Result<TcpStream>
        with reactor: &Reactor;
}
```

**Comment by uazu** [comment]
I guess you realize that there is not just tokio and async-std to consider. I want to implement this interface for Stakker, which is one-runtime-per-thread never-blocking actor-model runtime, with typically `mio` in the background taking care of waiting for lots of FDs/handles and waiting for the next timer expiry (although it could run on top of any other event loop). So "block_on" is problematic, and "spawn_blocking" too. Can I opt out of those interfaces? It will likely cause deadlocks.

If a crate that Stakker is hosting via this interface wants to spawn threads of its own, no problem, but apart from a cross-thread waker of some kind, the runtime can't have anything to do with what it's doing in other threads, since the entire runtime is non-Send/Sync. Okay, maybe not all crates will be able to exist within those limitations (i.e. maybe I can't host everything), but I think the vast majority of protocol stuff should be able to work at least.

Right now, the only stuff that's portable to Stakker is stuff that's very low-level, like `embedded-websocket`, i.e. something that does no I/O itself. Really a lot more crates could be written to be portable. So if some interface can be standardized, I will support it as far as is possible in a single-threaded runtime.

**Comment by NobodyXu** [comment]
> So "block_on" is problematic

block_on is now implemented as try_enter_block_on and try_exit_block_on https://github.com/nrc/portable-interoperable/issues/13#issuecomment-1186439456 .
You can just always return Err for these functions.

> and "spawn_blocking" too

Why would spawn_blocking cause deadlock?
It is usually implemented using a blocking thread-pool that is separated from the thread-pool for executing futures.

> the runtime can't have anything to do with what it's doing in other threads, since the entire runtime is non-Send/Sync

You can simply implement spawn using spawn_local.

And there will be an interface called spawn_auto that automatically spawns locally if the future does not implement Send so that users of the portable runtime doesn't have to implement these themselves.

**Comment by uazu** [comment]
> Why would spawn_blocking cause deadlock?  It is usually implemented using a blocking thread-pool that is separated from the thread-pool for executing futures.

Maybe I misunderstood the purpose of spawn_blocking. So it doesn't actually block. Yes, I can implement a thread-pool if that is what is required. I guess there are some additional complexities to handle, e.g. non-blocking I/O through this interface from these tasks needs a reactor in those threads (potentially?), also they may spawn normal tasks (back in the runtime thread?). The whole thing needs thinking through.

> You can simply implement spawn using spawn_local.  And there will be an interface called spawn_auto that automatically spawns locally if the future does not implement Send so that users of the portable runtime doesn't have to implement these themselves.

Okay. I don't think Stakker will be the only single-threaded runtime for Rust (I think there are others already), so ideally things should be arranged so that the base API supports single-threaded, and people coding against this interface will understand that they are limiting their portability if they use certain additional methods. If everyone uses 'spawn' by default then straight away a big chunk of the ecosystem is cut off. It's not just this kind of shard-per-thread runtime approach, but also embedded/etc where maybe there aren't even threads.

Whenever things have settled down a bit I will try to implement the interface.

**Comment by uazu** [comment]
> You can simply implement spawn using spawn_local

Okay, got it -- so 'spawn' won't be a problem.

**Comment by NobodyXu** [comment]
> Maybe I misunderstood the purpose of `spawn_blocking`.

It is for blocking io, e.g. doing file io on a polling runtime.

> I guess there are some additional complexities to handle, e.g. non-blocking I/O through this interface from these tasks needs a reactor in those threads (potentially?), also they may spawn normal tasks (back in the runtime thread?). The whole thing needs thinking through.

If your reactor does not implement `Send`, then the users will not be able to use the reactor in these blocking threads.
And I agree that the whole things still need to be worked on.

> If everyone uses 'spawn' by default then straight away a big chunk of the ecosystem is cut off.

I plan to let them use `spawn_auto`, which spawns on local if they do not implement `Send`.

> Okay, got it -- so 'spawn' won't be a problem.

You got it, `spawn` only enables the executor to run the tasks on any thread and move it between them, but you can still run them on the current thread.

**Comment by Noah-Kennedy** [comment]
I think the whole idea of spawn_auto is problematic. On many runtimes, you cannot spawn a local task from any context, so implicitly falling back to local spawning if the future is `!Send` is a significant footgun, and prevents this method from being used in the manner it seems intended, which is as a default usage pattern for libraries. Ultimately, we should try to make it so that libraries don't need to be aware of runtime details any more than absolutely necessary. There might be ways around this, and I have a few ideas here.

We do need a way of spawning that works for both runtimes where tasks are not sendable (as there are advantages to "isolated" thread-per-core runtimes), as well as work-stealing runtimes. @uazu brought up a really great point here, and we need to try and figure out a solution.

Give me some time, and I can get a proof-of-concept of my ideas hacked out later today.

**Comment by Noah-Kennedy** [comment]
Also, thinking on this more, we do not need `block_on`. This is an implementation detail that doesn't really serve a use case. We really just need a system for constructing the basic IO types, and doing spawning.

Also, another note since this ticket is getting heated at times:

Folks, let's please keep in mind that the problem we need to address is that library authors and others seeking to write systems that can work on multiple runtimes are unable to do so easily, due to the coupling to a runtime's spawn method primarily, and secondarily due to the need to couple to a runtime's specific IO types. We don't need to solve this by crafting a framework that tries to abstract over every implementation detail of every runtime. We just need to look at the minimum set of abstractions capable of making this easier, and make sure that these abstractions are broadly capable of being met by different classes of runtimes.

**Comment by joshtriplett** [comment]
@Noah-Kennedy We do need a std::task::block_on though.

**Comment by Noah-Kennedy** [comment]
@joshtriplett why? What use case does it help with?

**Comment by joshtriplett** [comment]
@Noah-Kennedy "I'm in synchronous code, and I need to await a future; I need some way to switch from sync mode to async mode." (The reverse of the `spawn_blocking` use case.)

**Comment by Noah-Kennedy** [comment]
How common is this for a library to need to do though? I cannot think of an actual time to use this.

**Comment by joshtriplett** [comment]
@Noah-Kennedy Also, sorry, I responded quickly to that one point without acknowledging and appreciating the rest of your message:

> Folks, let's please keep in mind that the problem we need to address is that library authors and others seeking to write systems that can work on multiple runtimes are unable to do so easily, due to the coupling to a runtime's spawn method primarily
...
> We don't need to solve this by crafting a framework that tries to abstract over every implementation detail of every runtime. We just need to look at the minimum set of abstractions capable of making this easier

I really appreciate and agree with this point, and thank you for expressing it.

**Comment by Noah-Kennedy** [comment]
And I am trying to weigh this against the fact that not all runtimes can do this (a lot of thread-per-core designs can't).

**Comment by joshtriplett** [comment]
> How common is this for a library to need to do though? I cannot think of an actual time to use this.

I've found it quite common in application code, extremely common in documentation/example/test code, and moderately common in library code (notably in library code that isn't 100% inside the async world).

I don't think either `spawn_blocking` or `block_on` can always-return-error, because that will break quite a bit of extant code. I would *much* rather have an interface that guarantees support for both of those, even if implementing them is easier on some runtimes than others.

> And I am trying to weigh this against the fact that not all runtimes can do this (a lot of thread-per-core designs can't).

As far as I can tell, any design *can*, but it may have to go through more steps to do so. I would like to understand better how a runtime *couldn't*, rather than just that it doesn't naturally fit into its model and requires an extra layer. Just as `spawn_blocking` likely requires a blocking thread pool, I would expect `block_on` to require some machinery that may be trivial for some runtimes but may be more complex for others.

I am hoping that the executor trait can serve as a backend for `std::task:block_on`, `std:task::spawn`, and `std::task::spawn_blocking`, the trio of which would serve as the primary interface to an executor. (I would also expect, based on our other conversations, to need some *minimal* reactor interface for saying "here are my waitable objects in case you want to wait on them in some way other than dedicated threads", though I personally think it might be appropriate to have a default implementation for those methods.)

**Comment by uazu** [comment]
> As far as I can tell, any design _can_, but it may have to go through more steps to do so. I would like to understand better how a runtime _couldn't_

In Stakker there are two contexts: outside the main loop, which is synchronous, and stuff called from the main loop (actor methods and future polling), which can never block. So how can I implement `block_on` if it is called from code called by the main loop? I cannot block at this point. Maybe if it was `block_on(...).await` then that would be non-blocking, but then that's just the same as `.await`.

Outside the main loop, I guess it could be used to start a standard main loop. Yes, that makes sense if that's going to be the standard way to start a new runtime. (I'll have to select what kind of main loop and what's pulled in with cargo features I guess, but that's my problem.)

Maybe the key is to have one interface for "running outside the runtime" which includes `block_on`, and another interface for "running inside the runtime" which doesn't.

**Comment by Noah-Kennedy** [comment]
@joshtriplett There are a _lot_ of different designs for what I am going to refer to as "isolated worker" thread-per-core runtimes, so I'm going to generalize a bit.

A lot of "isolated worker" thread-per-core designs try to basically model all of the worker threads as parallel systems, allowing for a lot less synchronization overall. This can be a very effective paradigm for server applications in conjunction with things like SO_REUSEPORT, and this is a very common pattern for custom runtimes written for a specific performance use cases or categories of workloads as a result.

Generally the setup process for these runtimes involves first spawning a bunch of threads pinned to different CPU cores, so `block_on` doesn't really work here; you can't enter the runtime context from outside the runtime to drive a particular future in the current thread, as the current thread you launch the runtime from isn't a part of the runtime context. Many of these designs also involve a factory pattern for initialization, as they tend to be custom-built for contexts where you are running a server on every worker, using SO_REUSEPORT to balance incoming connections between them. Spawning a task from within the runtime context involves allocating a new task on the worker and then inserting the task into the workers local queue (no synchronization required).

With these designs you _can_ support spawning from outside the runtime context quite easily via a global queue or other means (not going into this, there are lots of weird designs though). Generally, these designs are optimized under the assumption that spawning from outside the runtime is less common than spawning from inside the runtime, so the latter path is much faster.


These runtimes may seem a bit odd for those of us coming from async-std or tokio, but this is actually a well-established pattern that has considerable benefits for many applications, and which I suspect may become more common in the rust space due to the fact that it works _really_ well with io_uring. I'm trying to keep these designs in mind wrt our proposals here.

**Comment by joshtriplett** [comment]
@Noah-Kennedy If you can support spawning from outside the runtime context, in theory you could implement `block_on` as "spawn this future, make it signal a channel on completion, then synchronously wait on that channel". And `block_on` doesn't have to be the most *optimized* path, it just needs to work.

@uazu Generally speaking, some runtimes do have restrictions that you can't call `block_on` recursively from inside a future running on the runtime, only from code running any future or code in a `spawn_blocking` closure. We *could* document that as a restriction that applies to some runtime implementations.

**Comment by NobodyXu** [comment]
> @uazu Generally speaking, some runtimes do have restrictions that you can't call block_on recursively from inside a future running on the runtime, only from code running any future or code in a spawn_blocking closure. We could document that as a restriction that applies to some runtime implementations.

Maybe we can make this more explicit, by stating that spawning/block_on can only be used when any type that implements the `Executor` trait is available as a context parameter?

Then spawning/block_on can also make the

**Comment by joshtriplett** [comment]
@NobodyXu I don't think that would work; I would expect spawning (and the executor context) to be available everywhere, while block_on may be limited to contexts that aren't in a future (but the executor should still be available for spawning).

**Comment by NobodyXu** [comment]
> I don't think that would work; I would expect spawning (and the executor context) to be available everywhere

Yeah, I also expect the executor context to available on any future and inside the `spawn_blocking` closure.

For reactor, this is more complex as it might not implement `Send` if the rutime wants to have a reactor per thread design. 

> while block_on may be limited to contexts that aren't in a future (but the executor should still be available for spawning).

Not sure how to ensure we are not in a future though, I wish there is such mechanism in the language to express this because now several functions in `tokio` panic at runtime, which is IMO a footgun.

**Comment by joshtriplett** [comment]
I agree that it's a footgun, and it's nice when implementations avoid that footgun, but despite that, block_on is still useful and important.

**Comment by uazu** [comment]
> For reactor, this is more complex as it might not implement `Send` if the rutime wants to have a reactor per thread design.

From Stakker's point of view, if I implement block_on, and it is running inside the runtime, I will panic. The only other alternative would be to spawn the future, but that probably also breaks the user's expectations.

If I implement `spawn_blocking`, it is probably going to have to be under a completely different executor and reactor in that other thread. So I can follow whatever rules typically apply, like "allow block_on" or whatever.
 
> > while block_on may be limited to contexts that aren't in a future (but the executor should still be available for spawning).
> 
> Not sure how to ensure we are not in a future though, I wish there is such mechanism in the language to express this because now several functions in `tokio` panic at runtime, which is IMO a footgun.

There *is* such a mechanism, but you have to buy into it totally. This is how Stakker works. There are borrows that are passed down the stack, and if you don't have that borrow, then you don't have access to a whole bunch of stuff. This is all tied into QCell and friends (similar to GhostCell). It also means that Stakker only ever runs a (relatively) shallow stack before returning to the main loop, because if you get any deeper you lose the important borrows. I wrote this all up [HERE](https://uazu.github.io/stakker/d-intro.html). Unfortunately for futures, there is no way to pass the borrow through the `poll` call, because nobody thought of that. Or maybe it could be put into the Context but it needs someone high up to think it all through.  Generators I think *will* eventually have support for passing borrows into a resume, which become unavailable again at the next yield -- some day anyway.

**Comment by uazu** [comment]
> @uazu Generally speaking, some runtimes do have restrictions that you can't call `block_on` recursively from inside a future running on the runtime, only from code running outside any future or code in a `spawn_blocking` closure. We _could_ document that as a restriction that applies to some runtime implementations.

Yes, that sounds ideal. Perhaps we can have a way to express levels of support, e.g. Stakker might support level 2, and tokio/async_std level 4, or whatever. Then if a crate claims to only require level 1, then it's clear which runtimes it can run on. (Or maybe expressed as a required-feature list.)

**Comment by NobodyXu** [comment]
> Generators I think will eventually have support for passing borrows into a resume, which become unavailable again at the next yield -- some day anyway.

That will be an effect system, I'm not sure if rust wants to support that.

Perhaps we could work with the keyword generics wg to see if they have any idea to solve this?

P.S. @uazu Thanks for the links, using QhostCell sounds interesting and I will have a read at the links you provided.

**Comment by NobodyXu** [comment]
> Or maybe it could be put into the Context but it needs someone high up to think it all through. 

Note that this will not work, since users can come up with their own context.

For example, the crate futures has its own context type so passing the token in context just wouldn't work.

@uazu Perhaps you can also use the "with context" we planned to use to pass around executor/reactor to pass around your token?

**Comment by NobodyXu** [comment]
I think [this blog post](http://jblog.andbit.net/2022/12/28/context-reactor-hook/) has proposed a more elegant solution:

Instead of creating Reactor traits, it proposes extending Context to allow blocking callback to be added if not already set.

This would avoid the global variables, the Reactor traits and need of context parameters.

But if some library crates need to abstract over functionalities provided by async runtime reactors, they would still need the reactor traits.

**Comment by Noah-Kennedy** [comment]
I like the idea of extending context
